031-set-pipeline-step/proposal.md

```diff
@@ -0,0 +1,120 @@
+
+## `((vars))` support
+
+Additionally, we should support `vars` (as in `fly set-pipeline -y`) and `var_files` (i.e. `fly set-pipeline -l`):
+
+```yaml
+plan:
+- get: ci
+- set_pipeline: release
+  file: ci/pipelines/release.yml
+  vars: {release_version: 5.3}
+  var_files:
+  - ci/pipelines/vars/foo.yml
+```
+
+# Open Questions
+
+* Should we prevent pipelines configured via a `set_pipeline` step from being
+  updated manually through `fly set-pipeline`?
+
+  Or should we emit a warning?
+
+* Should we support glob expansion in `var_files`?
+
+  The `concourse-pipeline` resource supports this by just performing glob
+  expansion against its local filesystem. For the `set_pipeline` step, this is
+  a bit more challenging - there *is* no local filesystem. Would we have to
+  implement glob expansion in the Baggageclaim API or something? How easily
+  would this translate to other runtimes?
+
+* `set_pipeline: self`
+
+  Currently, the `foo` in `set_pipeline: foo` is the name of a pipeline to set.
+  A pipeline could technically update itself by configuring its own name in the
+  step, but pipeline configs aren't meant to contain their own name, as doing
+  so prevents the config from being re-used as a 'pipeline template'.
+
+  Are self-updating pipelines a feature that we want to explicitly support by
+  allowing the keyword `self` in place of the pipeline name?
+
+  The [Projects RFC][projects-rfc] outlines a more 'git-ops' style flow for
+  configuring pipelines, where instead of having pipelines self-update they are
+  all configured in one central place (the project's `plan:`).
+
+  Pros:
+
+  * Fairly straightforward semantics which seem to support a natural follow-up
+    question after learning about the `set_pipeline` step.
+  * Keyword use has precedent in `version: every`/`version: latest`/`inputs: all`.
+
+  Cons:
+
+  * Supporting both self-updating pipelines and projects could cause confusion
+    and fragmentation; it doesn't seem wise to have two competing approaches to
+    the same goal.
+  * Given that `self` isn't *critical* (it's easy to work around through
+    templating, i.e. `set_pipeline: ((name))`), is it worth the
+    risk/maintenance?
+
+
+# Answered Questions
+
+none
+
+
+# New Implications
+
+## Deprecating `concourse-pipeline` resource
+
+Deprecating the `concourse-pipeline` resource should be the primary goal.
+
+Some of the extended functionality of the resource will not be supported in the name of keeping the `set_pipeline` step design simple and easy to reason about.
+
+For example, the step should only ever configure one pipeline at a time - it should not support the `pipelines:` functionality for configuring a bunch at once.
```

Comment:
Is it possible to run set_pipeline step multiple times from a list of configuration to mimic the `pipelines:` functionality? I'm afraid the `concourse-pipeline` resource users can't swtich until they find a way to do so, in which case we can't really make the resource deprecated.

Reply:
This should probably be done via #29 - it could even be given a static list of pipeline configs to configure. If it's determined at runtime, that list could could be populated via the `load_var` step (#27) or come from a var source.